allow the opportunity to customize your logging factory

carlo-rtr opened the pull request.

jplock commented:
Should we also try and address https://github.com/dropwizard/dropwizard/pull/987 or keep that as a separate change?

carlo-rtr commented:
That's a good point, let's get that sorted. I'll add my 2 cents.

Btw, I probably should say that there was much debate on whether this should be configurable. See https://github.com/dropwizard/dropwizard/pull/567

My feeling is that we should only maintain the default logger. Making it flexible doesn't feel like it's going to add much over head on our maintenance. If someone is really hell bent on using xml configuration, they'll find a way.

arteam commented:
After #987 that issue need to rebased.

Also we need to think abount the need of bootstrapping logging in `Application`, if it's configured externally.

carlo-rtr commented:
I've rebased and added some new code.

* documented the bootstrap logging code
* perform bootstrap logging init in constructor of ```Application``` in favor of ```static``` block

carlo-rtr commented:
@arteam I missed the #987 comment. I'll wait until your code is back on master, then I'll pull it down and push the rebased version.

carlo-rtr commented:
@jplock @arteam should be all set for review

jplock commented:
LGTM!  I'll let @arteam review as well.

arteam commented on dropwizard-logging/src/main/java/io/dropwizard/logging/BootstrapLogging.java:
```diff
@@ -0,0 +1,67 @@
+ * strategy has been applied.
+ * <p/>
+ * N.B. The methods in this class have run once semantics,
+ * multiple calls are impodent
```
Nit: s/impodent/idempotent, if this word was meant.

arteam commented on dropwizard-logging/src/main/java/io/dropwizard/logging/BootstrapLogging.java:
```diff
@@ -0,0 +1,67 @@
+
+    @GuardedBy("boostrappingLock")
+    private static boolean bootstrapped = false;
+    private static Lock bootstrappingLock = new ReentrantLock();
```
Could we me make the lock final? I think it's better to always lock on final fields as a safety measure.
Moreover, this field is not mutated.

arteam commented on dropwizard-logging/src/main/java/io/dropwizard/logging/BootstrapLogging.java:
```diff
@@ -0,0 +1,67 @@
+ */
+public class BootstrapLogging {
+
+    @GuardedBy("boostrappingLock")
```
Nit: s/boostrappingLock/bootstrappingLock

arteam commented on dropwizard-logging/src/main/java/io/dropwizard/logging/BootstrapLogging.java:
```diff
@@ -0,0 +1,67 @@
+
+        bootstrappingLock.lock();
+        try {
+            if (!bootstrapped) {
```
Maybe better to revert the condition? If logging is bootstrapped, we can return early.
Advantage is a reduction of a big *if* block with identation.

arteam commented on dropwizard-logging/src/main/java/io/dropwizard/logging/LoggingUtil.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class LoggingUtil {
```
I wonder should we make this class a package protected or public... I mean, do we want its methods be visible for user logger factories or should they be closed as a part of internal Dropwizard logging mechanism? I'm more apt to to the latter. What is your opinion on this?

carlo-rtr replied:
I was thinking it should be public because so that people implementing their own logging factories have a way to get the LoggerContext in a thread safe way and have a utility for hijacking jdk logging.

arteam commented on dropwizard-logging/src/main/java/io/dropwizard/logging/LoggingUtil.java:
```diff
@@ -0,0 +1,71 @@
+
+    @GuardedBy("julHijackingLock")
+    private static boolean julHijacked = false;
+    private static Lock julHijackingLock = new ReentrantLock();
```
The lock could be final, because we don't mutate it.

arteam commented:
Apart from some nitpicks, this PR looks accurate to me.

carlo-rtr commented:
I really wish I could spell :)

carlo-rtr commented:
@arteam fixed the typos, made the locks final, and inverted the condition. Thanks for looking
